feat(test): add jest test running on npm test

heafalan opened the pull request:
Fixes #28

heafalan committed:
feat(test): add jest test running on npm test
Fixes #28

heafalan committed:
(test): install jest

heafalan commented:
I had to commit with the option "-n" to bypass 
because 'test' is not defined in the pre-commit.

julien-f commented:
You can enable the `jest` environment in the test file.

https://eslint.org/docs/user-guide/configuring#specifying-environments

julien-f commented on package.json:
```diff
@@ -24,8 +24,11 @@
   "scripts": {
     "start-server": "./packages/server/index.js",
-    "start-test": "./packages/test/index.js",
+    "start-test": "yarn jest packages/test/jest.test.js",
```
You don't need to prefix with `yarn` for the scripts.

julien-f commented on package.json:
```diff
@@ -24,8 +24,11 @@
     "postinstall": "cd packages/cli/ && npm ci && cd ../pages/ && npm ci && cd ../server/ && npm ci && cd ../test/ && npm ci"
+  },
+  "dependencies": {
+    "jest": "^23.6.0"
```
This should be a dev dependency.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,97 @@
+  await call("createEntry", { content: "content2" });
+  await call("createEntry", { name: "name3" });
+
+  var response = await call("listEntries");
```
Never use `var`, use `const` or `let` if the variable is re-assigned later.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,97 @@
+test("create entry", async () => {
+  await call("createEntry", { name: "name1", content: "content1" });
+  await call("createEntry", { content: "content2" });
+  await call("createEntry", { name: "name3" });
```
While we're at it, test the return values of all calls.

heafalan replied:
should I test the "true" values ​​returned by update and delete, or not?

julien-f replied:
No, that's probably unnecessary :slightly_smiling_face:

heafalan committed:
fix(test): test the return values of createEntry,
change jest dep in devDependency,
use const instead var

heafalan committed:
fix(test): add env jest in jest.test.js
+ remove "yarn" for testing with jest

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,103 @@
+
+const hrp = require("http-request-plus").default;
+const jrp = require("json-rpc-protocol");
+const assert = require("assert");
```
Because we are now using Jest, you can use the provided [`expect`](https://jestjs.io/docs/en/expect).

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,103 @@
+const assert = require("assert");
+
+async function call(method, params) {
+  return jrp.parse(
```
Use `parse.result`.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,103 @@
+}
+
+test("createe entry", async () => {
+  const id1 = await call("createEntry", { name: "name1", content: "content1" });
```
Test the result directly, instead of doing it later.

julien-f replied:
You can even remove the variable entirely.

heafalan committed:
fix(test): use expect instead of assert, and reorganize the tests

heafalan committed:
fix(test): use parse.result

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,93 @@
+        .readAll("utf-8")
+    );
+  } catch (error) {
+    return error;
```
If you return an error, how can you distinguish it from a result?

julien-f replied:
Still unresolved.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,93 @@
+  expect(response[0].name).toBe("name1");
+  expect(response[0].content).toBe("content1");
+  expect(response[1].content).toBe("content2");
+  expect(response[2].name).toBe("name3");
```
Take advantage of `expect(response).toEqual(...)` to simplify your tests.

heafalan committed:
fix(test): use toEqual instand of toBe

heafalan committed:
(test): 'toThrowError' not working yet

heafalan committed:
(test): remove await

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,93 @@
+    await call("createEntry", { name: "name1", content: "content1" })
+  ).toEqual(1);
+  expect(await call("createEntry", { content: "content2" })).toEqual(2);
+  expect(await call("createEntry", { name: "name3" })).toEqual(3);
```
Use `toBe` to compare scalar values.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,93 @@
+  expect(response[0].name).toEqual("name1");
+  expect(response[0].content).toEqual("content1");
+  expect(response[1].content).toEqual("content2");
+  expect(response[2].name).toEqual("name3");
```
```js
expect(response).toEqual([
  {
    id: 1,
    name: "name1",
    content: "content1",
  },
  {
    id: 2,
    content: "content2",
  },
  {
    id: 3,
    name: "name3",
  },
])
```

heafalan replied:
We can't do like that, there are two others parameters to precise in this case: "created" and "updated". But we don't know what is the value of its each time we run the tests.

julien-f replied:
Almost straight from [the documentation](https://jestjs.io/docs/en/expect#expectanyconstructor):

```js
expect(response).toEqual([
  {
    id: 1,
    name: "name1",
    content: "content1",
    created: expect.any(Number),
    updated: expect.any(Number),
  },
  {
    id: 2,
    content: "content2",
    created: expect.any(Number),
    updated: expect.any(Number),
  },
  {
    id: 3,
    name: "name3",
    created: expect.any(Number),
    updated: expect.any(Number),
  },
]);
```

heafalan committed:
(test): 'toThrow' still doesn't work

heafalan commented:
'toThrow' and 'toThrowError' doesn't work and I don't know why.
The function call didn't throw anything, when it has to.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,86 @@
+      name: "name0",
+      content: "content0",
+    })
+  ).toThrowError(new jrp.InvalidParameters(`could not find id 0`));
```
It does not work because an async function never throws, it returns an rejected promise, if you want to test the promise, you need to use [`.rejects`](https://devdocs.io/jest/expect#rejects).

julien-f replied:
Also, according to the documentation, there is no `toThrowError` matcher.

heafalan committed:
(tests): '.rejects.toThrow' doesn't work

heafalan commented:
The function call didn't throw anything.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,120 @@
+      name: "name0",
+      content: "content0",
+    })
+  ).rejects.toThrow();
```
`toThrow` does not seem to behave correctly when the value is not an error.

Here's my suggestion:
```js
const invertPromise = p =>
  p.then(
    value => {
      throw value;
    },
    reason => reason
  );

test("Error on delete: could not find id 0", async () => {
  await expect(
    invertPromise(
      call("updateEntry", {
        id: 0,
        name: "name0",
        content: "content0",
      })
    )
  ).resolves.toEqual({
    code: -32602,
    data: "could not find entry 0",
    message: "invalid parameter(s)",
  });
});
```

julien-f replied:
Also you may want to use `toMatchSnapshot` instead of `toEqual` :wink:

heafalan committed:
fix(test): add invertPromise() + use Snapshots

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,142 @@
+  );
+}
+
+const invertPromise = p =>
```
Add a small comment above explaining the use case for this.

julien-f replied:
This comment explains the *how*, but that can already be seen from the code itself, it would be more interesting to explain the *why*.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,142 @@
+test("create entry", async () => {
+  expect(
+    await call("createEntry", { name: "name1", content: "content1" })
+  ).toBe(1);
```
Change all your ids to be string, it's much more future-proof.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,142 @@
+  expect(
+    await call("createEntry", { name: "name1", content: "content1" })
+  ).toBe(1);
+  expect(await call("createEntry", { content: "content2" })).toBe(2);
```
Your tests should not be aware of how the ids are generated.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,142 @@
+      name: "name1",
+      content: "content1",
+      created: expect.any(Number),
+      updated: expect.any(Number),
```
We could probably test these more:

```js
const now = Date.now()

expect(e1.created).toBe(e2.updated)
expect(Math.abs(e1.created - now)).toBeLessThan(1e2)
```
```

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,142 @@
+      name: "name1_modified",
+      content: "content1_modified",
+      created: expect.any(Number),
+      updated: expect.any(Number),
```
We should test:
- `created` has not been changed
- the new value of `updated` is coherent

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,142 @@
+
+  const response = await call("listEntries");
+
+  expect(response[1].id).not.toBe(2);
```
This test could be improved to not depend on other entries:

```js
expect(response.include(entry => entry.id === 2)).toBe(false)
```

heafalan committed:
fix(test): edit the tests
- type of id : String
- hidden ids for all entries
- add tests for updated and created.

heafalan committed:
fix(test): replace the 'delete entry'

heafalan commented on packages/test/jest.test.js:
```diff
@@ -177,3 +169,15 @@ test("Error: method not found", async () => {
+  expect(response.includes(entry => entry.id === entry1.id)).toBe(false);
+  expect(response.includes(entry => entry.id === entry2.id)).toBe(false);
+  expect(response.includes(entry => entry.id === entry3.id)).toBe(false);
+});
```
I do not know if it's useful but you told me yesterday that we have to run the tests without relaunching the server (with jest --watch). I don't know if that is what you want me to do.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,183 @@
+  expect(entry3.created).toBe(entry3.updated);
+  expect(Math.abs(entry1.created - now)).toBeLessThan(1e2);
+  expect(Math.abs(entry2.created - now)).toBeLessThan(1e2);
+  expect(Math.abs(entry3.created - now)).toBeLessThan(1e2);
```
It would probably be nice to abstract this log in a `compareTimestamps()` helper.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,183 @@
+    entry3.created
+  );
+  expect(
+    Math.abs(response.find(entry => entry.id === entry1.id).updated - now)
```
Use variables instead of searching the entries multiple times.

julien-f replied:
Also it may be easier to use [`lodash/keyBy`](https://devdocs.io/lodash~4/index#keyBy) instead :slightly_smiling_face:

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,183 @@
+  ).toBeLessThan(1e2);
+});
+
+test("Error on delete: could not find id 0", async () => {
```
The message is not perfect: there is no longer a `0` id.

heafalan committed:
fix(test): edit the tests
- add compareTimestamps()
- comment invertPromise()
- use lodash

heafalan committed:
fix(test): use lodash/keyBy instead of lodash

julien-f committed:
Update jest.test.js

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,175 @@
+  expect(entries[id1].created).toBe(entries[id1].updated);
+  expect(entries[id2].created).toBe(entries[id2].updated);
+  expect(entries[id3].created).toBe(entries[id3].updated);
+  expect(compareTimestamps(entries[id1].created, now)).toBe(true);
```
IMHO it would be better to put the test (`expect`) itself in `compareTimestamps`:

- more code mutualization
- better error messages due to the use of `expect().toBeLessThan()`

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,175 @@
+
+const hrp = require("http-request-plus").default;
+const jrp = require("json-rpc-protocol");
+const _ = require("lodash/keyBy");
```
`_` is not a meaningful name, use `keyBy` instead.

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,175 @@
+
+  expect(response.includes(entry => entry.id === id1)).toBe(false);
+  expect(response.includes(entry => entry.id === id2)).toBe(false);
+  expect(response.includes(entry => entry.id === id3)).toBe(false);
```
This could also be improved with `keyBy`.

heafalan committed:
fix(test): mutualize code, edit compareTimestamps()...

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,175 @@
+  expect(entriesUpdated[id2].created).toBe(entries[id2].created);
+  expect(entriesUpdated[id3].created).toBe(entries[id3].created);
+
+  keyBy(response, function(entry) {
```
This is not what `keyBy` is intended for!

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,175 @@
+    expect(entry.id === id1 || entry.id === id2 || entry.id === id3).toBe(
+      false
+    );
+  });
```
This is not what `keyBy` is intended for.

```js
const entries = keyBy(await call("listEntries"), "id");

expect(id1 in entries).toBe(false);
```

heafalan committed:
fix(test): improve the use of 'keyBy'

heafalan commented on packages/test/jest.test.js:
```diff
@@ -111,9 +111,7 @@ test("update entry", async () => {
-  keyBy(response, function(entry) {
-    compareTimestamps(entry.updated, now);
-  });
+  keyBy(response, entry => compareTimestamps(entry.updated, now));
```
Is that what you expected or not ?

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,171 @@
+  expect(entriesUpdated[id2].created).toBe(entries[id2].created);
+  expect(entriesUpdated[id3].created).toBe(entries[id3].created);
+
+  keyBy(response, entry => compareTimestamps(entry.updated, now));
```
That's not the use case for `keyBy`.

julien-f replied:
You could use [`Array#forEach()`](https://devdocs.io/javascript/global_objects/array/foreach).

heafalan committed:
fix(test): using response.forEach()

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,171 @@
+
+  entries = keyBy(response, "id");
+
+  keyBy(response, function(entry) {
```
Why `keyBy`?

julien-f commented on packages/test/jest.test.js:
```diff
@@ -0,0 +1,171 @@
+  const entriesUpdated = keyBy(response, "id");
+  expect(entriesUpdated[id1].created).toBe(entries[id1].created);
+  expect(entriesUpdated[id2].created).toBe(entries[id2].created);
+  expect(entriesUpdated[id3].created).toBe(entries[id3].created);
```
Why don't you put these tests into the `forEach` as well?

heafalan committed:
fix(test)
